internal/langserver/handlers/command/module_list.go

```diff
@@ -0,0 +1,145 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/creachadair/jrpc2/code"
+	lsctx "github.com/hashicorp/terraform-ls/internal/context"
+	"github.com/hashicorp/terraform-ls/internal/langserver/cmd"
+	"github.com/hashicorp/terraform-ls/internal/terraform/datadir"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+const moduleListVersion = 0
+
+type moduleListResponse struct {
+	FormatVersion int               `json:"v"`
+	Modules       []terraformModule `json:"modules"`
+}
+
+type terraformModule struct {
+	Name             string            `json:"name"`
+	Path             string            `json:"path"`
+	Version          string            `json:"version"`
+	Type             string            `json:"type"`
+	DocLink          string            `json:"docklink"`
+	DependentModules []terraformModule `json:"depmodules"`
+}
+
+func ModuleListHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return nil, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return nil, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return nil, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	found, _ := mm.ModuleByPath(modPath)
+	if found.ModManifest == nil {
+		return nil, nil
+	}
+
+	list := parseModuleRecords(found.ModManifest.Records)
+
+	return moduleListResponse{
+		Modules:       list,
+		FormatVersion: moduleListVersion,
+	}, nil
+}
+
+func parseModuleRecords(records []datadir.ModuleRecord) []terraformModule {
+	modules := make(map[string]terraformModule)
+	for _, manifest := range records {
+		if manifest.Dir == "." {
```

Comment:
```suggestion
		if manifest.IsRoot() {
```
I _think_ this looks less magical here and you may not even need the comment below at that point.